src/rules/no-missing-atx-heading-space.js

```diff
@@ -10,17 +10,46 @@
 /**
  * @import { MarkdownRuleDefinition } from "../types.js";
  * @typedef {"missingSpace"} NoMissingAtxHeadingSpaceMessageIds
- * @typedef {[]} NoMissingAtxHeadingSpaceOptions
+ * @typedef {[{ checkClosedHeadings?: boolean }]} NoMissingAtxHeadingSpaceOptions
  * @typedef {MarkdownRuleDefinition<{ RuleOptions: NoMissingAtxHeadingSpaceOptions, MessageIds: NoMissingAtxHeadingSpaceMessageIds }>} NoMissingAtxHeadingSpaceRuleDefinition
  */
 
 //-----------------------------------------------------------------------------
 // Helpers
 //-----------------------------------------------------------------------------
 
-const headingPattern = /^(#{1,6})(?:[^# \t]|$)/u;
+const leadingAtxHeadingHashPattern = /^(#{1,6})(?:[^# \t]|$)/u;
+const trailingAtxHeadingHashPattern = /(?<![ \t])([ \t]*)(?<!\\)(#+)([ \t]*)$/u;
```

Comment:
Regex are error prone, we are recently seeing many bugs and misses with regex. I see the approach in https://github.com/eslint/markdown/pull/450 is regex free can you evaluate that if that would work for all the test cases you have added?